### PR TITLE
MAPSJS-2983 Don't create POI if image is missing in cache

### PR DESCRIPTION
**Pull request by nzjony**
The user can modify the cache so we need to check if the image exists before preparing the POI

Change-Id: I9a68e523e9e3aa8a815a53037063a9d83cd6a436
Signed-off-by: Jonathan Stichbury <2533428+nzjony@users.noreply.github.com>

Thank you for contributing to harp.gl!

Before requesting a pull request, please remember to check the following documents:
* [contribution guidelines](https://github.com/heremaps/harp.gl/blob/master/CONTRIBUTING.md)
* [coding style](https://github.com/heremaps/harp.gl/blob/master/CODINGSTYLE.md)

If you are adding new functionality we would highly appreciate if you can describe what is the capability you are adding and even better if you can add some examples. Please also remember to add tests for it.

# CI Check

Our bots will check whether your PR can be directly integrated into the mainline. We have some internal integration tests running on the background, our bots will inform you of the next steps and someone from our team will take a look and help if needed!

And please do not forget to sign-off your commit! You can read more about DCO [here](https://julien.ponge.org/blog/developer-certificate-of-origin-versus-contributor-license-agreements/). But, in short, you just need to use `git commit -s` or append `--signoff` when you are committing to the repo.

Happy contributing!

### CODE REVIEW

File: @here/harp-mapview/lib/poi/PoiRenderer.ts [hunk 1]
```diff
@@ -492,8 +485,12 @@ export class PoiRenderer {
         if (poiInfo === undefined) {
             return false;
         }
+        const { imageItem, imageTexture } = this.imageCached(poiInfo);
```
Comment (nzjony) [inline @here/harp-mapview/lib/poi/PoiRenderer.ts:488]: Here we prevent creation of the image isn't in the cache.

File: @here/harp-mapview/test/PoiRendererTest.ts [hunk 2]
```diff
@@ -97,4 +100,93 @@ describe("PoiRenderer", function () {
...
+            const testCache = new MapViewImageCache();
+            // This is the difference to the test above, we don't start loading, we expect the
+            // prepareRender function to do this.
+            addRandomImageToCache(testCache, testImageName, false);
+            const caches = [testCache];
+            const poiRenderer = new PoiRenderer(webGLRenderer, poiManager, caches);
+            const pointLabel = createPointLabel(testImageName);
+            const waitingForLoad = poiRenderer.prepareRender(pointLabel, mapEnv);
+            const result = await Promise.resolve(waitingForLoad);
+            expect(result).true;
```
Comment (nzjony) [inline @here/harp-mapview/test/PoiRendererTest.ts:169]: this is the test which required the change to return a Promise<boolean>

File: karma.options.js [hunk 3]
```diff
@@ -170,6 +170,7 @@ const options = function (isCoverage, isMapSdk, prefixDirectory) {
                 allowJs: true
             },
             coverageOptions: {
+                instrumentation: isCoverage ? true : false,
```
Comment (nzjony) [inline karma.options.js:173]: This is required to debug locally when running `yarn test-browser`

Comment (atomicsulfate) [reply karma.options.js:173]:
```suggestion
                instrumentation: isCoverage,
```

File: @here/harp-mapview/lib/poi/PoiRenderer.ts [hunk 4]
```diff
@@ -373,37 +373,21 @@ export class PoiBatchRegistry {
...
-            '${imageTexture?.image ?? imageTextureName}'`);
-        poiInfo.isValid = false;
-        if (missingTextureName.get(imageTextureName) === undefined) {
-            missingTextureName.set(imageTextureName, true);
-            logger.error(`preparePoi: No imageTexture with name '${imageTextureName}' found`);
+            return imageItem;
         }
     }
-    return imageItem;
+    return undefined;
```
Comment (atomicsulfate) [inline @here/harp-mapview/lib/poi/PoiRenderer.ts:390]: How do we catch actual errors, i.e. an image missing from cache that should actually be there, e.g. outdated atlas or style.

File: @here/harp-mapview/lib/poi/PoiRenderer.ts [hunk 5]
```diff
@@ -486,14 +470,29 @@ export class PoiRenderer {
...
             return false;
         }
+        const { imageItem, imageTexture } = this.imageCached(poiInfo.imageTextureName);
+        if (!imageItem) {
+            // Set the POI to be invalid, so it will not be rendered, nor will it be re-checked in a
+            // future request. The item must exist (but not necessarily loaded in the cache before
+            // the POI can be initialized). We don't log an error, because the API of the
+            // MapViewImageCache allows the user to delete an image from the cache. The geometry
+            // creation is handled using some task queue which means that in some cases, a POI entry
+            // may be removed before the geometry is created. See queueGeometryCreation, which
```
Comment (atomicsulfate) [inline @here/harp-mapview/lib/poi/PoiRenderer.ts:488]: I would avoid to make explicit references to code outside this file that is not even part of the API. They become obsolete quickly.

File: @here/harp-mapview/lib/poi/PoiRenderer.ts [hunk 6]
```diff
@@ -486,14 +470,29 @@ export class PoiRenderer {
...
+     * or a Promise if the image has to be loaded first. This happens when the user explicitly
+     * doesn't load the image when it is put into the cache.
      */
-    prepareRender(pointLabel: TextElement, env: Env): boolean {
+    prepareRender(pointLabel: TextElement, env: Env): boolean | Promise<boolean> {
         const poiInfo = pointLabel.poiInfo;
         if (poiInfo === undefined) {
             return false;
         }
+        const { imageItem, imageTexture } = this.imageCached(poiInfo.imageTextureName);
```
Comment (atomicsulfate) [inline @here/harp-mapview/lib/poi/PoiRenderer.ts:481]: There's no need to do this if the poiInfo.buffer is already created (see the if below). Why did you move this out of preparePoi?

File: @here/harp-mapview/lib/poi/PoiRenderer.ts [hunk 7]
```diff
@@ -486,14 +470,29 @@ export class PoiRenderer {
...
+        if (!imageItem) {
+            // Set the POI to be invalid, so it will not be rendered, nor will it be re-checked in a
+            // future request. The item must exist (but not necessarily loaded in the cache before
+            // the POI can be initialized). We don't log an error, because the API of the
+            // MapViewImageCache allows the user to delete an image from the cache. The geometry
+            // creation is handled using some task queue which means that in some cases, a POI entry
+            // may be removed before the geometry is created. See queueGeometryCreation, which
+            // appends some item to the list. In such cases, the POI will just be abandoned and the
+            // geometry will be recreated with the new ImageItem.
+            poiInfo.isValid = false;
```
Comment (atomicsulfate) [inline @here/harp-mapview/lib/poi/PoiRenderer.ts:491]: In the Venues case, the image is not in cache cause the theme is not yet loaded. Here we'll mark pois referencing to images that are not yet loaded as invalid, and they'll never be rendered.

Do you plan to fix that on JSAPI side by waiting for the theme to load?

File: @here/harp-mapview/lib/poi/PoiRenderer.ts [hunk 8]
```diff
@@ -539,7 +538,10 @@ export class PoiRenderer {
 
         if (opacity > 0) {
             if (!poiInfo.buffer) {
-                this.preparePoi(poiInfo.textElement, env);
+                const { imageItem, imageTexture } = this.imageCached(poiInfo.imageTextureName);
```
Comment (atomicsulfate) [inline @here/harp-mapview/lib/poi/PoiRenderer.ts:541]: ditto, why do we need to move this here? Furthermore, now it's in two places, prepareRender() and addPoi()

File: @here/harp-mapview/test/PoiRendererTest.ts [hunk 9]
```diff
@@ -97,4 +103,107 @@ describe("PoiRenderer", function () {
...
+            // This is false because the image is still being loaded in the background, notice the
+            // true flag passed to the call to testCache.addImage.
+            expect(waitingForLoad).false;
+
+            // Promise.resolve used to convert the type `ImageItem | Promise<ImageItem>` to
+            // `Promise<ImageItem>`.
+            await Promise.resolve(testImageItem);
+
+            const imageLoaded = poiRenderer.prepareRender(pointLabel, mapEnv);
+            expect(imageLoaded).true;
```
Comment (atomicsulfate) [inline @here/harp-mapview/test/PoiRendererTest.ts:161]: Why does this pass? Looking at the code, it semes that preparePoi would return false at line 675, cause poinInfo.isValid would have been set to false in the previous call.

File: @here/harp-mapview/test/PoiRendererTest.ts [hunk 10]
```diff
@@ -97,4 +103,107 @@ describe("PoiRenderer", function () {
...
+            const waitingForLoad = poiRenderer.prepareRender(pointLabel, mapEnv);
+            const result = await Promise.resolve(waitingForLoad);
+            expect(result).true;
+
+            // Check that a second attempt to prepareRender succeeds.
+            const imageLoadedAgain = poiRenderer.prepareRender(pointLabel, mapEnv);
+            expect(imageLoadedAgain).true;
+        });
+
+        it("poi is invalid when not in cache, adding to cache doesn't help", async function () {
```
Comment (atomicsulfate) [inline @here/harp-mapview/test/PoiRendererTest.ts:187]: Ditto, this test is saying that the Venues issue is not solved (an image not found because the theme is still being loaded). Will we solve that on JSAPI side?

File: @here/harp-mapview/test/PoiRendererTest.ts (hunk 2 above)
Comment (atomicsulfate) [reply @here/harp-mapview/test/PoiRendererTest.ts:180]: If you added the promise just for testing, there's another option that would leave the production code intact. You can wait for the async tasks to be executed doing:

`const wait = (ms: number = 0) => new Promise(res => setTimeout(res, ms));

await wait();
`
It's used in TileGeometryLoaderTest, TileGeometryCreatorTest and DecodedTilesHelperTest.

**Review by atomicsulfate — CHANGES_REQUESTED** [review]
review done

File: @here/harp-mapview/lib/poi/PoiRenderer.ts (hunk 4 above)
Comment (nzjony) [reply @here/harp-mapview/lib/poi/PoiRenderer.ts:390]: The problem is, with the current architecture / API, we don't know when this situations occurs.

The current task queue geometry creation makes the assumption that the image cache is fixed until the geometry is created, but the user can modify the cache before the task has executed.

I break that assumption and say we can create geometry for something which doesn't exist in the image cache, but the POI is invalid, and it is up to the user to make the tiles dirty and re-request them (such is the responsibility of using a custom cache).

File: @here/harp-mapview/lib/poi/PoiRenderer.ts (hunk 5 above)
Comment (nzjony) [reply @here/harp-mapview/lib/poi/PoiRenderer.ts:488]: I'll remove it.

File: @here/harp-mapview/lib/poi/PoiRenderer.ts (hunk 7 above)
Comment (nzjony) [reply @here/harp-mapview/lib/poi/PoiRenderer.ts:491]: I don't think it is needed, if I check the JSAPI side, the `setTheme`, then `await addDataSource`, which internally anyway `await`s on the theme.

File: @here/harp-mapview/lib/poi/PoiRenderer.ts (hunk 6 above)
Comment (nzjony) [reply @here/harp-mapview/lib/poi/PoiRenderer.ts:481]: I wanted to centralise the point where the poiInfo.isValid = false is set, it is just set in this place, as opposed to being set in two places previously.

And it simplifies the `preparePoi` a bit.

But good point about the fact that I don't need this if the buffer is already created, I'll think about this.

File: @here/harp-mapview/test/PoiRendererTest.ts (hunk 10 above)
Comment (nzjony) [reply @here/harp-mapview/test/PoiRendererTest.ts:187]: There are two things, an item not being in the cache and an item in the cache not being loaded.

The first thing will cause the test to fail, the second won't, because it will be loaded.

File: @here/harp-mapview/test/PoiRendererTest.ts (hunk 2 above)
Comment (nzjony) [reply @here/harp-mapview/test/PoiRendererTest.ts:180]: As discussed, I tried this, but it doesn't work, so will resolve.

File: @here/harp-mapview/lib/poi/PoiRenderer.ts (hunk 6 above)
Comment (nzjony) [reply @here/harp-mapview/lib/poi/PoiRenderer.ts]: I moved it, so will resolve.

File: @here/harp-mapview/test/PoiRendererTest.ts (hunk 2 above)
Comment (atomicsulfate) [reply @here/harp-mapview/test/PoiRendererTest.ts:180]: @nzjony You missed this comment.

File: @here/harp-mapview/lib/poi/PoiRenderer.ts (hunk 4 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/poi/PoiRenderer.ts:390]: I was thinking about the cache used to store theme images. The user has no control over that one. In theory we should always find images referenced in the style. 
It's quite possible that at some point we use an outdated atlas we an updated style, specially now that we don't generate it automatically. Without even an error in the log, the missing resources might go unnoticed for a long time.

I think it's better to keep logging the errors. We can make them warnings instead of errors, maybe also say in the message that it might just be a transient error.

**Review by atomicsulfate — COMMENTED** [review]
still some questions open
